src/lib/profiles/security/WeaveCertProvisioning.cpp

```diff
@@ -0,0 +1,767 @@
+        {
+            client->mState = kState_Idle;
+
+            client->SendGetCertificateRequest();
+        }
+
+        break;
+
+    case Binding::kEvent_PrepareFailed:
+
+        // If binding preparation fails and the client is still in the PreparingBinding state...
+        if (client->mState == kState_PreparingBinding)
+        {
+            // Clean-up the request state and enter Idle state.
+            client->HandleRequestDone();
+
+            // Deliver a CommunicationError API event to the application.
+            client->DeliverCommunicationError(inParam.PrepareFailed.Reason);
+        }
+
+        break;
+
+    default:
+        Binding::DefaultEventHandler(appState, eventType, inParam, outParam);
+    }
+}
+
+/**
+ * Called by the Weave messaging layer when a response is received on an GetCertificate exchange.
+ */
+void WeaveCertProvEngine::HandleResponse(ExchangeContext * ec, const IPPacketInfo * pktInfo, const WeaveMessageInfo * msgInfo, uint32_t profileId, uint8_t msgType, PacketBuffer * msgBuf)
+{
+    WeaveCertProvEngine * client = (WeaveCertProvEngine *)ec->AppState;
+
+    VerifyOrDie(client->mState == kState_RequestInProgress);
+    VerifyOrDie(client->mEC == ec);
+
+    // Process status report message from the CA service.
+    if (profileId == kWeaveProfile_Common && msgType == Common::kMsgType_StatusReport)
+    {
+        StatusReport statusReport;
+        WEAVE_ERROR parseErr = StatusReport::parse(msgBuf, statusReport);
+
+        // Free the received message buffer so that it can be reused to initiate additional communication.
+        PacketBuffer::Free(msgBuf);
+        msgBuf = NULL;
+
+        SuccessOrExit(parseErr);
+
+        VerifyOrExit(statusReport.mProfileId == kWeaveProfile_Security, /* no-op */);
+
+        InEventParam inParam;
+        OutEventParam outParam;
+        inParam.Clear();
+        outParam.Clear();
+        inParam.Source = client;
+
+        if (statusReport.mStatusCode == Security::kStatusCode_NoNewCertRequired)
+        {
+            inParam.ResponseReceived.ReplaceCert = false;
+
+            // Deliver a ResponseReceived API event to the application.
+            client->mEventCallback(client->AppState, kEvent_ResponseReceived, inParam, outParam);
+        }
+        else
+        {
+            inParam.CommunicationError.Reason = WEAVE_ERROR_STATUS_REPORT_RECEIVED;
+            inParam.CommunicationError.RcvdStatusReport = &statusReport;
+
+            // Deliver a CommunicationError API event to the application.
+            client->mEventCallback(client->AppState, kEvent_CommunicationError, inParam, outParam);
+        }
+    }
+
+    // Process GetCertificateResponse message from the CA service.
+    else if (profileId == kWeaveProfile_Security && msgType == kMsgType_GetCertificateResponse)
+    {
+        // If the request is now complete, clear the request state and enter Idle state.
+        client->HandleRequestDone();
```

Comment:
I think this is broken.  HandleRequestDone() always aborts the EC.  In this case, however, aborting the will result in the ACK never being sent to the service.   A simple solution would be to expressly close (and NULL) the EC here before calling HandleRequestDone().

Reply:
done